### PR TITLE
Support expanding locations in rustc_env and build_script_env #2

### LINKED ISSUE #79: Including a genfile at compile time is not possible
Minimal repro case with a potential (but ugly) fix:

https://github.com/bazelbuild/rules_rust/compare/master...SirVer:00_include_genfiles?expand=1

The linked branch does the following changes:

1. Add a [genrule](https://github.com/bazelbuild/rules_rust/compare/master...SirVer:00_include_genfiles?expand=1#diff-e85472ad14a7cde9a8510d9beae332c5R10) call that trivially creates a file containing a string.
2. The goal is to `include_str!` this file in `hello_world.rs`. Code for this is [here](https://github.com/bazelbuild/rules_rust/compare/master...SirVer:00_include_genfiles?expand=1#diff-0208a81a165de01670ea086341ddbd59R19). For this to ever work, the genfile is made a `data` dependency of the `rust_binary`.
3. I could not pull this off without exposing bazel's `ctx.bin_dir` and `ctx.genfiles_dir` as environment variables. This is still ugly, since it includes "external", so the file needs to have the context that it is embedded as an external workspace.

Ideally, I would just want to `include_str!("message.string");` or ` include_str!("../message.string");`, but both do not currently work. Is there a better way or solution to this issue? Should my repro case become a patch?

### LINKED ISSUE #454: Bazel build with prost-build build.rs fails with "No such file or directory"
I created a demo repository:

https://github.com/ttiurani/bazel-prost-build-bug

to showcase an issue where the [standard `prost-build` example](https://docs.rs/prost-build/0.6.1/prost_build/) works fine with a `cargo build` but running the same with `cargo raze`'d Bazel repository with `bazel build demo` fails with:

```
ERROR: /private/var/tmp/_bazel_ttiurani/0c87cdffcf4741baaa498655bde07714/external/raze__demo_prost_build__0_0_1/BUILD.bazel:36:19: CargoBuildScriptRun external/raze__demo_prost_build__0_0_1/demo_prost_build_build_script.out_dir failed (Exit 1) cargo_build_script_runner failed: error executing command bazel-out/darwin-opt-exec-2B5CBBC6/bin/external/io_bazel_rules_rust/cargo/cargo_build_script_runner/cargo_build_script_runner ... (remaining 8 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', external/raze__demo_prost_build__0_0_1/cargo/build.rs:2:66
stack backtrace:
   0:        0x10cac9a6e - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h24bb64d98a7e25d6
   1:        0x10caeaf9c - core::fmt::write::h8fdc9cddb01cd8b2
   2:        0x10cac6bd9 - std::io::Write::write_fmt::hcc3030013983bab6
   3:        0x10cacbc35 - std::panicking::default_hook::{{closure}}::h95817712c5ff0736
   4:        0x10cacb972 - std::panicking::default_hook::h34e085f4e0b1062d
   5:        0x10cacc195 - std::panicking::rust_panic_with_hook::haf571858f996ac45
   6:        0x10cacbd62 - rust_begin_unwind
   7:        0x10caf37ef - core::panicking::panic_fmt::h11676ba6a846d9f4
   8:        0x10caf36f5 - core::option::expect_none_failed::h312e48bacb63d8ec
   9:        0x10ca81c1b - demo_prost_build_build_script_script_::main::h399dfdb84fd54273
  10:        0x10ca82986 - std::rt::lang_start::{{closure}}::h0a0a8f1aceef40a7
  11:        0x10cacc45f - std::rt::lang_start_internal::hcd84a36052901671
  12:        0x10ca81c69 - main
Error: "Build script process failed with exit code 101"
```

Where the build.rs is:

```
fn main() {
    prost_build::compile_protos(&["src/items.proto"], &["src/"]).unwrap();
}
```

The error message is quite non-descriptive so was unable to trace this further myself. Note that I'd very much like to use the bundled `protoc` binary and not an external one to make my build easily repeatable in CI environments.

Any ideas what might be wrong? I'm using the latest rules_rust from this morning.

### LINKED ISSUE #459: Can't use `include_str!()` macro with Bazel-generated data files
## The Problem

The `data` attribute appears on each of the core Rust rules (e.g. `rust_binary`). The documentation for `data` says

> This attribute can be used to specify any data files that are embedded into the library, such as via the `include_str!` macro.

I have found that `include_str()!` works only for "source" files (i.e. files included in the Bazel source tree). Unfortunately, the `include_str!()` macro does **not** work if the data files are generated by some other Bazel target.

## Example

To demonstrate this, I've [added a small test to my `rules_rust` fork](https://github.com/dwtj/rules_rust/commit/b67ce49f0bc81735f4e083da560eb02fd6911fbf).

In this test, one data file is a source file and one data file is generated by a `genrule`. The contents of both files are embedded in the `rust_test` via the `include_str!()` macro.

On my linux system, this `rust_test` fails to compile with the following `rustc` error:

```
error: couldn't read test/generated_data/include_str/generated_data.txt: No such file or directory (os error 2)
 --> test/generated_data/include_str/include_str.rs:9:13
  |
9 |             include_str!("generated_data.txt").trim()
  |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error
```

Notice that `rustc` has no problem finding `source_data.txt`.

I believe that the problem can be seen by inspecting this `rustc` action's sandbox. On my Linux system, if I run the test with `bazel test --sandbox_debug`, and then change directory to `<bazel_workspace_output_base>/sandbox/linux-sandbox/<action_id>/execroot/io_bazel_rules_rust`, the root of the problem becomes clear: the library and the source data are sibling files

```
./test/generated_data/include_str/include_str.rs
./test/generated_data/include_str/source_data.txt
```

but the generated file is not

```
./bazel-out/k8-fastbuild/bin/test/generated_data/include_str/generated_data.txt
```

So, the generated data file is an input to the `rustc` Bazel action, but it isn't anywhere that the `rustc` executable can find it. Thus, a compiler error occurs.

## Temporary Workaround

In my project where I actually want this feature I have an awful-dreadful-ugly hack to overcome this problem: I add a symlink from my source tree pointing to where I expect the generated file should be under the Bazel workspace's output base; then I add both the generated file and its symlink to the `data` attribute.

## A Very Rough Idea for a Solution

I have a little bit of experience writing Bazel rules that also needed to deal with this problem. My solution [there](https://github.com/dwtj/dwtj_rules_hugo/blob/8617d3455cf8cc9787aced9ea2a86045efefeb7d/hugo/private/hugo_website/defs.bzl#L57) was for each target to add various `ctx.actions.symlink()` actions to place both generated files and source files into a new directory under their logical paths before calling the tool.

### PR DESCRIPTION
**Pull request by dae**
This is #461 squashed into fewer commits - please see that PR for previous discussion.

This potentially closes #459, closes #454, and closes #79.

The docs seem to indicate there's precedent for this in rules_cc:
https://docs.bazel.build/versions/master/be/make-variables.html#predefined_label_variables

### CODE REVIEW

**Review by illicitonion — APPROVED** [review]
Works for me!

One implementation detail worth specifically calling out: This expands to an absolute, rather than relative, path for cargo build scripts, because we don't run them in the execroot. I think this is fine, but it's a bit weird.

I will leave a bit of time for others to chime in before merging :)

File: cargo/cargo_build_script.bzl [hunk 1]
```diff
@@ -4,6 +4,19 @@ load("@io_bazel_rules_rust//rust:private/rustc.bzl", "BuildInfo", "DepInfo", "ge
 load("@io_bazel_rules_rust//rust:private/utils.bzl", "find_toolchain")
 load("@io_bazel_rules_rust//rust:rust.bzl", "rust_binary")
 
+def _expand_location(ctx, env, data):
+    if env.startswith("$(execpath ") or env.startswith("$(location "):
```
[the remark this answers is not in the document]
Comment (dae) [reply cargo/cargo_build_script.bzl]: I suspect non-prefixed usage won't be common, but it was easy enough to change. Also added a note about it being an absolute path to the docstring.

**Review by illicitonion — APPROVED** [review]
Thanks!

**Comment by dae** [comment]
Latest push just corrects some $(execroot) references in the docs to $(execpath)

File: cargo/cargo_build_script.bzl [hunk 2]
```diff
@@ -220,10 +238,19 @@ def cargo_build_script(
...
-        # Environment variables passed during build.rs execution
-        build_script_env = {"CARGO_PKG_VERSION": "0.1.2"},
+        # Optional environment variables passed during build.rs compilation
+        rustc_env = {
+           "CARGO_PKG_VERSION": "0.1.2",
+        },
+        # Optional environment variables passed during build.rs execution.
+        # Note that as the build script's working directory is not execroot,
+        # execpath/location will return an absolute path, instead of a relative
+        # one.
```
Comment (mfarrugi) [inline cargo/cargo_build_script.bzl:248]: Does this work under remote-execution? Is it tested by the ci rbe_ubuntu1604 target?

Comment (illicitonion) [reply cargo/cargo_build_script.bzl:248]: Yes, it should work under remote execution - the expansion is done in the cargo_build_script runner within the action.

The build script could opt to do bad things with the path passing it to dependees (though we have some logic in cargo_build_script to try to normalise that too), but within the action this is valid.

**Comment by dae** [comment]
Rebased on latest master; if there's anything else that needs changing, please let me know.

**Pull request merged**